lib/handler/connect.c

```diff
@@ -0,0 +1,189 @@
+    } server_name;
+    struct {
+        struct st_server_address_t list[MAX_CONNECT_RETRIES];
+        size_t size;
+        size_t next;
+    } server_addresses;
+    h2o_timer_t timeout;
+};
+
+static void start_connect(struct st_connect_request_t *creq);
+
+static void on_error(struct st_connect_request_t *creq, const char *errstr)
+{
+    h2o_timer_unlink(&creq->timeout);
+    h2o_send_error_502(creq->src_req, "Gateway Error", errstr, 0);
+}
+
+static void on_timeout(h2o_timer_t *entry)
+{
+    struct st_connect_request_t *creq = H2O_STRUCT_FROM_MEMBER(struct st_connect_request_t, timeout, entry);
+    on_error(creq, h2o_httpclient_error_io_timeout);
+}
+
+static void on_connect(h2o_socket_t *sock, const char *err)
+{
+    struct st_connect_request_t *creq = sock->data;
+
+    if (err) {
+        if (creq->server_addresses.next == creq->server_addresses.size) {
+            on_error(creq, err);
+            return;
+        }
+        start_connect(creq);
+        return;
+    }
+
+    h2o_timer_unlink(&creq->timeout);
+    h2o_req_t *req = creq->src_req;
+    uint64_t timeout = creq->handler->config.tunnel.timeout;
+    sock->data = NULL;
+    creq->sock = NULL;
+    req->res.status = 200;
+
+    h2o_tunnel_t *tunnel = h2o_tunnel_create_from_socket(sock);
+    req->establish_tunnel(req, tunnel, timeout);
+}
+
+static void on_generator_dispose(void *_self)
+{
+    struct st_connect_request_t *creq = _self;
+
+    if (creq->getaddr_req != NULL) {
+        h2o_hostinfo_getaddr_cancel(creq->getaddr_req);
+        creq->getaddr_req = NULL;
+    }
+
+    if (creq->sock != NULL)
+        h2o_socket_close(creq->sock);
+}
+
+static void store_server_addresses(struct st_connect_request_t *creq, struct addrinfo *res)
+{
+    /* copy first entries in the response; ordering of addresses being returned by `getaddrinfo` is respected, as ordinary clients
+     * (incl. forward proxy) are not expected to distribute the load among the addresses being returned. */
+    do {
+        struct st_server_address_t *dst = creq->server_addresses.list + creq->server_addresses.size++;
+        dst->sa = h2o_mem_alloc_pool_aligned(&creq->src_req->pool, H2O_ALIGNOF(struct sockaddr), res->ai_addrlen);
+        memcpy(dst->sa, res->ai_addr, res->ai_addrlen);
+        dst->salen = res->ai_addrlen;
+    } while (creq->server_addresses.size < PTLS_ELEMENTSOF(creq->server_addresses.list) && (res = res->ai_next) != NULL);
+}
+
+static void on_getaddr(h2o_hostinfo_getaddr_req_t *getaddr_req, const char *errstr, struct addrinfo *res, void *_creq)
+{
+    struct st_connect_request_t *creq = _creq;
+
+    assert(getaddr_req == creq->getaddr_req);
+    creq->getaddr_req = NULL;
+
+    if (errstr != NULL) {
+        on_error(creq, errstr);
+        return;
+    }
+
+    store_server_addresses(creq, res);
+    start_connect(creq);
+}
+
+static void start_connect(struct st_connect_request_t *creq)
+{
+    /* repeat connect(pop_front(address_list)) until we run out of the list */
```

Comment:
Do I understand correctly that each address failure would take 300s? That's an inordinately long time if so, but maybe I'm wrong.

Reply:
Nice catch!

I'm not sure if we need to have a smaller timeout (or if we should provide a knob for setting different values for DNS lookup, connection establishment, idle timeout).

Though, unless we introduce a separate per-address connection establishment timeout (either by providing a knob or hard-coding it), there's no point in trying multiple addresses. That's because `h2o_socket_connect` would take too long that the client would abort the CONNECT request (or the idle timeout would fire).

Reply:
As of 601609f, the CONNECT handler refers to `proxy.timeout.connect` and `proxy.timeout.io` for respective timeout values. `proxy.tunnel.timeout` has been removed.